packages/fuels-core/src/code_gen/functions_gen.rs

```diff
@@ -65,18 +66,76 @@ pub fn expand_function(
     })
 }
 
+/// Generate the function (as a `TokenStream`) used to encode into script data the arguments to the
+/// `main` function of a script.
+pub fn generate_script_main_function(
+    main_function_abi: &ABIFunction,
+    types: &HashMap<usize, TypeDeclaration>,
+) -> Result<TokenStream, Error> {
+    if main_function_abi.name != "main" {
+        return Err(Error::InvalidData(
+            "Script `main` function name can not be different from `main`".into(),
+        ));
+    }
+
+    let output_type_resolved = resolve_fn_output_type(main_function_abi, types)?;
+    let output_params = single_param_type_call(&output_type_resolved);
+    let output_type: TokenStream = output_type_resolved.into();
+
+    let args = function_arguments(main_function_abi, types)?;
+    if args.iter().filter(|c| c.field_type.uses_vectors()).count() != 0 {
+        return Err(Error::CompilationError(
+            "Script main function contains a vector in its argument types. This currently isn't \
+            supported."
+                .to_string(),
+        ));
+    }
+    let arg_names = args.iter().map(|component| &component.field_name);
+
+    let arg_declarations = args.iter().map(|component| {
+        let name = &component.field_name;
+        let field_type: TokenStream = (&component.field_type).into();
+        quote! { #name: #field_type }
+    });
+
+    let doc = expand_doc(
+        "Encode the arguments provided so they can be passed as argument to the script's main \
+        function",
+    );
+
+    let name = safe_ident("main");
+
+    Ok(quote! {
+        #doc
+        pub async fn #name(&self #(,#arg_declarations)*) -> Result<#output_type, SDKError> {
+            let provider = self.wallet.get_provider().expect("Provider not set up").clone();
+            let arg_name_tokens = [#(#arg_names.into_token()),*];
+            let script_data = ABIEncoder::encode(&arg_name_tokens)?.resolve(0);
+            let mut receipts = run_script_binary(self.binary_filepath.as_str(),
+                                                    Some(TxParameters::default()),
+                                                    Some(provider),
+                                                    Some(script_data)
+            )
+            .await?;
+            let output_tokens = get_decoded_output(#output_params, &mut receipts)?;
+            let output = #output_type::from_token(output_tokens)?;
```

Comment:
You're going to have the same issue here as you did earlier with the param type call since from token is also an associated fn.

You can either do the same logic or do the following workaround and leave rust to figure out the type:
```rust
    return Tokenizable::from_token(output_token);
```


Also, rename `output_tokens` to `output_token` since there can only be one.

Also, since this is the second time we're noticing stuff related to generics, add a test with generic script fn args?

Make sure there isn't a third thing we've missed.